### PR TITLE
DRAFT : Od

### PR DESCRIPTION
**Pull request by SupperZum**

### CODE REVIEW

File: text/src/od.rs [hunk 1]
```diff
@@ -0,0 +1,718 @@
...
+        }
+        if self.octal_words {
+            self.type_strings = vec!["o2".to_string()];
+        }
+        if self.signed_decimal_words {
+            self.type_strings = vec!["d2".to_string()];
+        }
+        if self.hex_words {
+            self.type_strings = vec!["x2".to_string()];
+        }
```
Comment (0x07C0) [inline text/src/od.rs:118]: Is it supposed to overwrite previous flag?

File: text/src/od.rs [hunk 2]
```diff
@@ -0,0 +1,718 @@
...
+
+    let base_value = if number.starts_with("0x") || number.starts_with("0X") {
+        u64::from_str_radix(&number[2..], 16)?
+    } else if number.starts_with('0') && number.len() > 1 {
+        u64::from_str_radix(&number[1..], 8)?
+    } else {
+        number.parse()?
+    };
+
+    Ok(base_value * multiplier)
```
Comment (0x07C0) [inline text/src/od.rs:187]: Very similar to `parse_count` fn. Possibly use generics to specify to which type we should parse (usize/u64/etc)

File: text/src/od.rs [hunk 3]
```diff
@@ -0,0 +1,718 @@
...
+            _ => print!("{:07} ", offset),
+        }
+    } else {
+        print!("{:07} ", offset);
+    }
+    Ok(())
+}
+
+/// Get a mapping of byte values to their named character representations.
+fn get_named_chars() -> HashMap<u8, &'static str> {
```
Comment (0x07C0) [inline text/src/od.rs:579]: Could've been a compile-time look-up table. Just accept the byte and match over it. Why the need for a map?

File: text/src/od.rs [hunk 4]
```diff
@@ -0,0 +1,718 @@
...
+    {
+        let mut buffer = Vec::new();
+        io::stdin().lock().read_to_end(&mut buffer)?;
+        buffer
+    } else {
+        let mut bufs: Vec<u8> = vec![];
+        for file in &args.files {
+            let mut buffer = Vec::new();
+            let mut file = std::fs::File::open(file)?;
+            file.read_to_end(&mut buffer)?;
```
Comment (0x07C0) [inline text/src/od.rs:653]: Will die on large files

File: text/src/od.rs [hunk 5]
```diff
@@ -0,0 +1,718 @@
...
+                let mut chars = type_string.chars();
+                let type_char = chars.next().unwrap();
+                let num_bytes: usize = chars.as_str().parse().unwrap_or(match type_char {
+                    'd' | 'u' | 'o' | 'x' => 2,
+                    'f' => 4,
+                    _ => 1,
+                });
+
+                let chunks = local_buf.chunks(num_bytes);
+                match type_char {
```
Comment (0x07C0) [inline text/src/od.rs:339]: There's definitely a better way to do this. On possible idea is to have a generic function depending on the type_char. Have a trait that describes printing and multiple structs (a,c,u,d,x,o,f,default) that implement this trait.

Function body will have skip checks and printing itself, and the trait method would return `current` value.

Of course you can do it your way, just make sure that it's simpler without a ton of repetition.

File: text/src/od.rs [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+                };
+
+                if previously == current && !config.verbose {
+                    print!("* ");
+                    continue;
+                }
+
+                print!("{}", current);
+                previously = current;
+            }
```
Comment (0x07C0) [inline text/src/od.rs:312]: Same as with the other comment about a ton of duplication.

File: text/src/od.rs [hunk 7]
```diff
@@ -0,0 +1,718 @@
...
+            for byte in local_buf {
+                let current = format!("{:03o} ", byte);
+                if previously == current && !config.verbose {
+                    print!("* ");
+                    continue;
+                }
+
+                print!("{}", current);
+                previously = current;
+            }
```
Comment (0x07C0) [inline text/src/od.rs:325]: Same as with the other comment about a ton of duplication.

**Review by 0x07C0 — CHANGES_REQUESTED** [review]

File: text/src/od.rs (hunk 1 above)
Comment (SupperZum) [reply text/src/od.rs:118]: There can be only 1, so if it overwrites it, validate_args will return an error.

File: text/src/od.rs [hunk 8]
```diff
@@ -0,0 +1,771 @@
...
+                    'c' => {
+                        process_formatter(&CFormatter, local_buf, config.verbose);
+                    }
+                    'u' => {
+                        if !(num_bytes == 1 || num_bytes == 2 || num_bytes == 4 || num_bytes == 8) {
+                            return Err(Box::new(Error::new(
+                                ErrorKind::Other,
+                                format!("invalid type string `u{}`", num_bytes),
+                            )));
+                        }
```
Comment (0x07C0) [inline text/src/od.rs:338]: Maybe needs a validator to remove duplication? Not critical, just a bit dirty

File: text/src/od.rs [hunk 9]
```diff
@@ -0,0 +1,771 @@
...
+        skip_bytes(&mut stdin, &mut bytes_to_skip)?;
+        if bytes_to_skip == 0 {
+            stdin.read_to_end(&mut buffer)?;
+        }
+    } else {
+        for file in &args.files {
+            let mut file = File::open(file)?;
+            skip_bytes(&mut file, &mut bytes_to_skip)?;
+            if bytes_to_skip == 0 {
+                file.read_to_end(&mut buffer)?;
```
Comment (0x07C0) [inline text/src/od.rs:719]: You've ignored comment about this issue. Reading the whole file to the buffer will cause insufficient memory and crash the app (best case scenario), or even make the system unusable (definitely not something we'd want)

File: text/tests/integration.rs [hunk 10]
```diff
@@ -2205,3 +2534,189 @@ mod diff_tests {
...
+    }
+
+    #[test]
+    fn test_od_4() {
+        od_test(
+            &[],
+            "Hello, World!",
+            "0000000 110 145 154 * 157 054 040 127 157 162 154 144 041 \n0000013 ",
+        );
+    }
```
Comment (0x07C0) [inline text/tests/integration.rs:2568]: > If no output type is specified, the default output shall be as if -t oS had been specified.

https://pubs.opengroup.org/onlinepubs/9699919799/utilities/od.html

**Review by 0x07C0 — CHANGES_REQUESTED** [review]

**Review by 0x07C0 — APPROVED** [review]

**Pull request closed without merging**